NF: uses hash set/map which take size as init instead of capacity

Arthur-Milchior opened the pull request:
Fixes #8337

david-allison-1 reviewed (changes requested):
```
  symbol:   class HashUtil
  location: package com.ichi2.utils
/home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/main/java/com/ichi2/themes/HtmlColors.java:3: error: cannot find symbol
import com.ichi2.utils.HashUtil;
```

Added a few places where the `HashUtil` is unused - this could indicate a problem, if it's fine, mark as resolved.

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/libanki/Models.java:
```diff
@@ -1112,7 +1111,7 @@ public static boolean emptyStandardCard(JSONObject tmpl, Set<String> nonEmptyFie
      * Cache of getNamesOfFieldsContainingCloze
      * Computing hash of string is costly. However, hash is cashed in the string object, so this virtually ensure that
-     * given a card type, we don't need to recompute the hash.
+     * given a card type, we don't need to recompute the HashUtil.
```
I used script to do some replacement. But this one, I fail to see how it was catched. Make no sens

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/libanki/template/ParsedNode.java:
```diff
@@ -1,15 +1,13 @@
 package com.ichi2.libanki.template;
 
 import android.content.Context;
-import android.text.TextUtils;
```
No need for this change

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/libanki/AnkiPackageExporter.java:
```diff
@@ -22,6 +22,7 @@
 import com.ichi2.anki.CollectionHelper;
 import com.ichi2.anki.R;
 import com.ichi2.anki.exception.ImportExportException;
+import com.ichi2.utils.HashUtil;
```
remove unused header
```suggestion
```

Arthur-Milchior replied:
no, I never saw a codacy email

mikehardy reviewed (changes requested):
Seems fine to me - there is one unused import (doesn't codacy still send you email with new issues when it finds them? I got email on my last PR that had a codacy report showing I'd created issues and it pointed out 2 unused imports I forgot...)

Also needs a rebase / de-conflict

mikehardy commented:
@Arthur-Milchior Something went terribly wrong in the rebase and now HashUtil, which is pretty important to this PR, is not in the change, causing predictable havoc...

mikehardy commented:
Hey @Arthur-Milchior - can you patch this one up? Then we can get it in before moving on to the other PRs you are posting :)

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/utils/HashUtil.java:
```diff
@@ -0,0 +1,21 @@
+        return new HashSet<T>(capacity(size));
+    }
+    public static <T, U> HashMap<T, U> HashMapInit(int size) {
+        return new HashMap<T, U>(capacity(size));
```
I spent some time to read about reflection but I don't understand how I could check that a function is not called. I don't see a way to change a method's code

david-allison-1 replied:
You can't (easily) - but you can read the internal state of the object to see if threshold is changed

Arthur-Milchior commented:
Sorry, I missed the first message. Too many things right now

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.java:
```diff
@@ -2034,7 +2034,7 @@ private void updateToolbar() {
     @NonNull
     private ArrayList<CustomToolbarButton> getToolbarButtons() {
-        Set<String> set = AnkiDroidApp.getSharedPrefs(this).getStringSet("note_editor_custom_buttons", new HashSet<>(0));
+        Set<String> set = AnkiDroidApp.getSharedPrefs(this).getStringSet("note_editor_custom_buttons", HashUtil.HashSetInit(0));
```
If we're getting into microoptimisations, we shouldn't be allocating the HashSet at all

Fine though

david-allison-1 reviewed (approved):
LGTM, had a play in the debugger and it seems good. A unit test would be very nice

Arthur-Milchior commented:
I rebased and added what I thought would be correct test. But it tells that some field I see in the source code do not exists. So I assume it uses a different implementation. Don't see how to test here. Help wanted

david-allison-1 commented:
> I rebased and added what I thought would be correct test. But it tells that some field I see in the source code do not exists. So I assume it uses a different implementation. Don't see how to test here. Help wanted

(On my phone). It's probably JaCoCo. Try filtering on `!field.isSynthetic`

david-allison-1 commented on AnkiDroid/src/test/java/com/ichi2/utils/HashUtilTest.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    @Test
+    public void NoReThreshold() throws NoSuchFieldException, IllegalAccessException {
```
```suggestion
    public void thresholdIsNotModifiedAfterCreationWithFixedSizeCollection() throws NoSuchFieldException, IllegalAccessException {
```

david-allison-1 commented on AnkiDroid/src/test/java/com/ichi2/utils/HashUtilTest.java:
```diff
@@ -0,0 +1,40 @@
+
+    /** Threshold of a set. Not a setter*/
+    private int setThreshold(HashSet<?> set) throws NoSuchFieldException, IllegalAccessException {
+        return mapThreshold((HashMap) HashMap.class.getField("map").get(set));
```
```suggestion
        Field map = HashMap.class.getDeclaredField("map");
        map.setAccessible(true);
        return mapThreshold((HashMap) map.get(set));
```

david-allison-1 commented on AnkiDroid/src/test/java/com/ichi2/utils/HashUtilTest.java:
```diff
@@ -0,0 +1,40 @@
+
+    /** Size of the table used to save a map.*/
+    private int mapThreshold(HashMap<?, ?> map) throws NoSuchFieldException, IllegalAccessException {
+        Field thresholdField = HashMap.class.getField("threshold");
+        return  thresholdField.getInt(map);
```
```suggestion
        Field thresholdField = HashMap.class.getDeclaredField("threshold");
        thresholdField.setAccessible(true);
        return thresholdField.getInt(map);
    }
```

david-allison-1 reviewed (commented):
Try this

david-allison-1 commented:
Bad suggestion - apologies - split my code into 3 for ease of understanding

david-allison-1 commented on AnkiDroid/src/test/java/com/ichi2/utils/HashUtilTest.java:
```diff
@@ -0,0 +1,44 @@
+        Field thresholdField = HashMap.class.getDeclaredField("threshold");
+        thresholdField.setAccessible(true);
+        return thresholdField.getInt(map);
+    }
```
Reminder to self: don't use suggestions for nontrivial fixes
```suggestion
```

Arthur-Milchior commented:
May I just let it to you, will probably get easier.

david-allison-1 commented:
Sorry this took so long

Given the CI output, let's get this in without tests:

```
> Task :AnkiDroid:testDebugUnitTest
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.ichi2.utils.HashUtilTest (file:/home/runner/work/Anki-Android/Anki-Android/AnkiDroid/build/intermediates/javac/debugUnitTest/classes/) to field java.util.HashMap.threshold
WARNING: Please consider reporting this to the maintainers of com.ichi2.utils.HashUtilTest
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release

com.ichi2.utils.HashUtilTest > thresholdIsNotModifiedAfterCreationWithFixedSizeCollection FAILED
    java.lang.AssertionError: expected:<16> but was:<12>
        at org.junit.Assert.fail(Assert.java:89)
        at org.junit.Assert.failNotEquals(Assert.java:835)
        at org.junit.Assert.assertEquals(Assert.java:647)
        at org.junit.Assert.assertEquals(Assert.java:633)
        at com.ichi2.utils.HashUtilTest.thresholdIsNotModifiedAfterCreationWithFixedSizeCollection(HashUtilTest.java:37)
```

I'm happy it's working from experimentation

david-allison-1 reviewed (changes requested):
adding this review flag: 

requested change: remove the tests given my comment above

Arthur-Milchior commented:
Done

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/utils/HashUtil.java:
```diff
@@ -0,0 +1,21 @@
+package com.ichi2.utils;
```
Missing copyright header check got you! :-)

```

/home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/main/java/com/ichi2/utils/HashUtil.java:1: Error: All files in AnkiDroid must contain a GPLv3-compatible copyright header [MissingCopyrightHeader from com.ichi2.anki.lint]
...
``

mikehardy reviewed (changes requested):
I was all ready to hit the tempting + fun-to-hit merge button but we need a copyright header, then I think it's merge-able

Arthur-Milchior commented:
done

david-allison-1 reviewed (approved):
LGTM, could do with a "Convert to Kotlin" for the new file, but that's not standardised yet

mikehardy reviewed (approved):
No need to keep messing around with this one - it took a conflict from a merge I just did but the code looks great on review now, I just went ahead and fixed the conflict if it passes CI I will squash-merge - thanks Arthur!
